src/geometry_aux.cpp

```diff
@@ -181,12 +183,15 @@ neighbor_lists()
 //==============================================================================
 
 void
-prepare_distribcell(int32_t* filter_cell_list, int n)
+prepare_distribcell()
 {
-  // Read the list of cells contained in distribcell filters from Fortran.
+  // Find all cells listed in a DistribcellFilter.
   std::unordered_set<int32_t> distribcells;
-  for (int i = 0; i < n; i++) {
-    distribcells.insert(filter_cell_list[i]);
+  for (auto* filt : tally_filters) {
+    if (filt->type() == "distribcell") {
+      auto* distrib_filt = static_cast<DistribcellFilter*>(filt);
```

Comment:
`dynamic_cast` would make more sense here

Reply:
I didn't make this change yet because I want to look more into static_cast and how it actually works.  My first instinct is that using `dynamic_cast` for RTTI is a bit of a hack and that we could get weird behavior so maybe the correct thing to do is implement our own RTTI through something like the `type()` function.  I'll get back to you on this.

Reply:
Conceptually `static_cast` and `dynamic_cast` do the same thing on a pointer, which is to say they don't do much at all. The purpose of `dynamic_cast` is to safely downcast from a base class pointer to a derived class pointer which is exactly what is being done here. If anything, you could get weird behavior with `static_cast` because there is no check that the type you're casting to is actually derived from what you started with. `dynamic_cast` guards against that by returning a null pointer.

Reply:
See bullet 2 [here](https://en.cppreference.com/w/cpp/language/static_cast)

Reply:
Right, I get that `static_cast` doesn't do any runtime checks, but what I'm pondering is if we should use our own runtime check system in this case because we want something more limited than `dynamic_cast`.

I'm worried that we could run into trouble with chains of inheritance like `Filter`->`MeshFilter`->`MeshSurfaceFilter`.  Imagine I'm writing `openmc_mesh_filter_set_mesh` and I forget about the `MeshSurfaceFilter`.  Checking `filt->type() == "mesh"` will fail for a `MeshSurfaceFilter`, but `dynamic_cast<MeshFilter*>(filt)` will succeed which can lead me to set the wrong value for `n_bins_`.  We can protect ourselves from errors like this if we're strict about using virtual setters which hide the `n_bins_` details from the C-API functions, but I still feel like this kind of typing error is easy to make.

Reply:
I've switched all of these to `dynamic_cast`.  We just need to be careful about making sure everything is defined as a virtual function when it needs to be.  (I guess if we run into errors like I described earlier, it means we messed up somewhere along the line with properly mapping abstractions to classes.)